### PR TITLE
[Table] RTL  and Table filters combo box bug fixes

### LINKED ISSUE #446: RTL: Table: Lack of right alignment and other issues
**Columns**
- Column content doesn't right-align
- Date header doesn't right-align, but others do
image

**Search**
- Icon doesn't switch to right of text entry
image

**Column controls**
Text in column elements doesn't right-align
image

**Filters**
Dropdown option menu doesn't right-align
image

**Expanding rows**
- Content in expanded row doesn't right-align
- Labels don't shift to the right of values
image

**Overflow menu**
- No padding btw text and icon
image

**Multi-select**
- Actions don't move to left, items selected doesn't move to right (in the table toolbar, those things do switch)
- Divider line btw Cancel and other actions disappears

image

**Page control**
- Items per page element doesn't right-align and overlaps with other elements
- Divider lines aren't in the right place
- Next / Back arrows point wrong way
image

**Dumb table & minitable**
Content and headers don't right align
image

**Nested table rows**
- Indent in nested rows offsets to the right, should indent to the left
image

**No data & custom empty state**
- "no data" texts aligns left, not right
image

**Horizontal scroll - full width**
- the end of the right-most column can't be horizontally scrolled to, which means you'll lose pieces of primary information when they are right-aligned.  The reader will lost the beginning of everything. (this is actually a problem in LTR as well. submitted as issue #448)
image

### LINKED ISSUE #1108: [Table] z-index issue with dropdowns in table filters
z-index issue with dropdowns in table filters and the table below. Dropdown goes behind the table and you are unable to see the dropdown options

### Steps to reproduce:
- go to [Table story](https://carbon-addons-iot-react.com/?path=/story/watson-iot-table--stateful-example-with-expansion-and-column-resize)
- click on filters
- click on select menu
- see table over dropdown

image

### PR DESCRIPTION
**Pull request by davidicus**
Closes #446 closes #1108

**Summary**

- several RTL issues with various pieces of the data Table outlined in the comments below
- bug fix for combo box that are used in our data table filter

**Acceptance Test (how to verify the PR)**

- Check areas of the table outlined in the comments below. Most defects can be found with [this](https://deploy-preview-1112--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-table--with-no-data) story
- open up knob panel and activate RTL mode
- You should see all areas behave properly in RTL mode

### CODE REVIEW

**Comment by davidicus** [comment]
> **Columns**
> 
> - [x] Column content doesn't right-align
> - [x] Date header doesn't right-align, but others do
> 
> image
> 
> **Search**
> 
>  - [x] Icon doesn't switch to right of text entry
>   image
> 
> **Column controls**
> - [x]Text in column elements doesn't right-align
> image
> 
> **Filters**
> - [x] Dropdown option menu doesn't right-align
> image
> 
> **Expanding rows**
> 
>  - [x] Content in expanded row doesn't right-align
> * ~Labels don't shift to the right of values~ - this is user provided content
>   image
> 
> **Overflow menu**
> 
> - [x] No padding btw text and icon
>   image
> 
> **Multi-select**
> 
> - [x] Actions don't move to left, items selected doesn't move to right (in the table toolbar, those things do switch)
> - [x] Divider line btw Cancel and other actions disappears
> 
> image
> 
> **Page control**
> 
> - [x] Items per page element doesn't right-align and overlaps with other elements
> - [x] Divider lines aren't in the right place
> - [x] Next / Back arrows point wrong way
>   image
> 
> **Dumb table & minitable**
> ~Content and headers don't right align~ - resolved in other PR
> image
> 
> **Nested table rows**
> 
> - [x] Indent in nested rows offsets to the right, should indent to the left
>   image
> 
> **No data & custom empty state**
> 
> * ~"no data" texts aligns left, not right~ - this design has changed and defect is no longer present
>   image
> 
> **Horizontal scroll - full width**
> 
> *~the end of the right-most column can't be horizontally scrolled to, which means you'll lose pieces of primary information when they are right-aligned.  The reader will lost the beginning of everything. (this is actually a problem in LTR as well. submitted as issue #448)~ - resolved in other PR
>   image

**Comment by bjornalm** [comment]
Removing `overflow:hidden` from `bx--table-header-label` will create an issue with truncated text and sort icons in the column headers 
image

**Comment by davidicus** [comment]
> Removing `overflow:hidden` from `bx--table-header-label` will create an issue with truncated text and sort icons in the column headers
> image
> 
> We should only remove it for the filter row

@bjornalm thanks. Only applying to filter rows now

**Comment by davidicus** [comment]
> I'm not 100% if these are in scope here, but some things I've noticed. Screenshots in the expands below:
> 
> Overflow menu opens the wrong way, causes overflow
> image
> 
> Pagination looks goofy - seems like some elements should be moved around to flow RTL properly
> image

@tay1orjones The overflow one is definitely in need of fix. For the second item. All BIDI instruction i have seen on UI is to mirror most things which is what you see there. The only additional difference is the forward and backward buttons needed to be flipped which they have been.

**Comment by tay1orjones** [comment]
Sounds good. With that fix and merge conflicts resolved, this looks good to me to merge.

I don't see pal-tracking issues linked for the two parent issues.

File: src/components/Table/TableBody/TableBodyRow/TableBodyRow.story.jsx [hunk 1]
```diff
@@ -54,11 +54,23 @@ storiesOf('Watson IoT/TableBodyRow', module)
       isExpanded={boolean('isExpanded', false)}
```
Comment (sstone2423) [inline src/components/Table/TableBody/TableBodyRow/TableBodyRow.story.jsx:54]: Do we still need these stories?

**Comment by davidicus** [comment]
> Looks good. A few comments that may or may not be issues
> 
> 1. The pagination text seems to be mix-matched. Considering this is static text, we should probably go ahead and align them correctly? For example, `Of 10 pages 1`
> 
> Screen Shot 2020-05-01 at 9 45 37 AM
> 
> 1. The left side actions aren't aligned
> 
> Screen Shot 2020-05-01 at 9 45 21 AM
> 
> 1. Is it correct that the `!` should be on the left here?
> 
> Screen Shot 2020-05-01 at 9 44 14 AM

For your first and 3rd item I saw this but do not know enough about RTL languages to know if that is what it would look like. Since the browser does this and does not come from our styling I was under the impression that browser knew better how it should be. There may be some things that the english language just doesn't look right being in RTL. For example spanish will have their punctuation for questions or and exclamations in the front and the back of a sentence. 

2nd item should be addressed

File: src/components/Table/TableBody/TableBodyRow/TableBodyRow.story.jsx (hunk 1 above)
Comment (davidicus) [reply src/components/Table/TableBody/TableBodyRow/TableBodyRow.story.jsx:54]: This PR was just to address RTL issues. I agree stories should be pared down but that work should happen in another PR linked to another issue.

**Comment by tay1orjones** [comment]
I've opened a new issue for the row actions overflow menu issue that will be solved via your new hook @davidicus 

https://github.com/IBM/carbon-addons-iot-react/issues/1197

I think we just need to resolve the conflicts here and we'll be good to merge 👍

File: src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx [hunk 2]
```diff
@@ -126,8 +121,27 @@ const onClick = (e, id, action, onApplyRowAction) => {
 class RowActionsCell extends React.Component {
   state = {
     isOpen: false,
+    ltr: true,
   };
 
+  componentDidMount() {
+    // this is latest changes
+    console.log('load:', document.dir);
```
Comment (scottdickerson) [inline src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx:129]: we should pull out the console.logs

File: src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx [hunk 3]
```diff
@@ -126,8 +121,25 @@ const onClick = (e, id, action, onApplyRowAction) => {
...
+    }
+  }
+
+  componentDidUpdate(prevProp, prevState) {
+    // this is latest changes
+    const { ltr } = this.state;
+    const isLtr = document.dir === 'ltr';
+    if (prevState.ltr !== isLtr) {
+      this.setState(state => ({ ltr: !state.ltr }));
+    }
```
Comment (tay1orjones) [inline src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx:140]: @davidicus CI is failing here

```
/home/travis/build/IBM/carbon-addons-iot-react/src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx
  136:13  error  'ltr' is assigned a value but never used  no-unused-vars
```

should it be this?

```suggestion
    const { ltr } = this.state;
    const isLtr = document.dir === 'ltr';
    if (prevState.ltr !== isLtr) {
      this.setState(state => ({ ltr: !ltr }));
    }
```

or were you intending to use prevState.ltr instead?

File: src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx [hunk 4]
```diff
@@ -126,8 +121,25 @@ const onClick = (e, id, action, onApplyRowAction) => {
...
 
+  componentDidMount() {
+    // this is latest changes
+    if (document.dir === 'rtl') {
+      this.setState(state => ({ ltr: !state.ltr }));
+    }
+  }
+
+  componentDidUpdate(prevProp, prevState) {
+    // this is latest changes
```
Comment (davidicus) [inline src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx:135]:
```suggestion
    
```

File: src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx [hunk 5]
```diff
@@ -126,8 +121,25 @@ const onClick = (e, id, action, onApplyRowAction) => {
...
+    const { ltr } = this.state;
```
Comment (davidicus) [inline src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx:136]:
```suggestion

```

File: src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx [hunk 6]
```diff
@@ -126,8 +121,25 @@ const onClick = (e, id, action, onApplyRowAction) => {
 class RowActionsCell extends React.Component {
   state = {
     isOpen: false,
+    ltr: true,
   };
 
+  componentDidMount() {
+    // this is latest changes
```
Comment (davidicus) [inline src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx:128]:
```suggestion

```

File: src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx (hunk 3 above)
Comment (davidicus) [reply src/components/Table/TableBody/RowActionsCell/RowActionsCell.jsx]: I think the linter is complaining about the unused variable. Removing and will see if travis passes.

**Pull request merged**